### PR TITLE
Making ruby executables work + ruby_rspec macro + more

### PR DESCRIPTION
**Pull request by kigster**
### What’s Changed?

This is a PR that improves and refactors the bundle install command, its steps, build file generation and making installed gems and their binaries available for consumption by the ruby libraries, binaries and tests in the BUILD files.

We also spent considerable amount of time testing various ways of running rspec tests and were able to add three examples to simple script that show three ways of running rspec against a library: 1. Using RubyOpt with -rrspec/autorun, 2. Specifying rspec binary as `@bundle//:bin/rspec`, and finally 3. By using a new macro and a rule `ruby_rspec`, which provides sensible defaults and implements (2) in starlark.

We moved a lot of hard coded definitions into the constants file. Most defaults are now specified there.

We added a facility to declare a used attribute as deprecated and log it to the output, 

We renamed `bundle_install` into `ruby_bundle` to be more consistent and more meaningful in this context, as `ruby_bundle` not only installs gems but makes them accessible later individually.

We refactored starlark code implementation of ruby_bundle to be modular making it easy to invoke bundler with different commands during the installation.

#### Other notable changes 

 * We can't disable gems because Bundler fails to work at all

 * Changing the default of `bundle_install` to `ruby_bundle`, but keeping `bundle_install` as a backwards compatible alias.

 * Now calling `bundle config `to set bundler configuration instead of passing command line options, which have been deprecated and are now causing errors with Bundler 2.1.2

### CODE REVIEW

File: ruby/private/bundle/bundle.bzl [hunk 1]
```diff
@@ -1,123 +1,162 @@
...
+        "deployment": "'true'",
+        "standalone": "'true'",
+        "frozen": "'true'",
+        "without": "development,test",
+        "path": "lib",
+        "jobs": "20",
+    }
 
-    bundler = Label("//:bundler/exe/bundler")
+    for option, value in [(option, value) for option, value in bundler_config.items()]:
```
Comment (kigster) [inline ruby/private/bundle/bundle.bzl]: This needs reverse order

File: ruby/private/bundle/bundle.bzl [hunk 2]
```diff
@@ -1,123 +1,162 @@
...
-    # Install the Gems into the workspace
+        result = run_bundler(runtime_ctx, args)
+        if result.return_code:
+            message = "Failed to set bundle config {} to {}: {}".format(
+                option,
+                value,
+                result.stderr,
+            )
+            fail(message)
+
```
Comment (kigster) [inline ruby/private/bundle/bundle.bzl]: The reason it’s only installing pure ruby gems is only due to the script we use which does unpack only, not instal.

File: ruby/private/bundle/bundle.bzl [hunk 3]
```diff
@@ -1,123 +1,171 @@
...
+    }
 
-    install_bundler(
-        ctx,
-        interpreter_path,
-        "install_bundler.rb",
-        "bundler",
-        ctx.attr.version,
-    )
+    for option, value in [(option, value) for option, value in bundler_config.items()]:
```
Comment (ClareCat) [inline ruby/private/bundle/bundle.bzl]: What is the intention with this?

File: examples/simple_script/BUILD.bazel [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        "//lib:foo",
+        "@bundle//:awesome_print",
+        "@bundle//:colored2",
+    ],
+)
+
+# This is an example of the RSpec definition that uses autorun
+# and points to spec_helper as the main spec file. It specifies
+# which specs to run using the args.
+ruby_test(
```
Comment (ClareCat) [inline examples/simple_script/BUILD.bazel]: When do these test targets get run in CI?

**Comment by ClareCat** [comment]
Thinking about this PR more this needs to be split into a couple smaller PRs that have less surface area. There is so much going on here that it's hard to keep track of the different parts. Any issues that arise from this PR will result in everything needing to be reverted instead of just the single piece that's an issue.

File: ruby/private/bundle/bundle.bzl (hunk 3 above)
Comment (kigster) [reply ruby/private/bundle/bundle.bzl]: As partially described in the comment above, this runs bundler to set the configuration flags in that directory — basically bundler is deprecating command line flags in favor of setting up local config before running bundle install. 

So if today you can:
```
bundle install --deployment --standalone --path=vendor/bundle --jobs=20 --binstubs=bin
```
in the future you will have to:
```
bundle config --local deployment
bundle config --local standalone
bundle config --local path=vendor/bundle
bundle binstubs --path=bin
```
This is what this code does.

**Pull request closed without merging**